### PR TITLE
Upgraded to VS2017 and added NetStandard 2.0

### PR DESCRIPTION
**Pull request by snakefoot**
Also made some minor changes:

- No longer catches all exceptions, but gives them to NLog (NLog uses the exception to notify wrappers about failures)
- Added more logging about failures, so they can be monitored in NLog-InternalLogger.
- Fixed the nuget-package so it correctly uses MsgPack.Cli ver. 0.8.0 (Instead of ver. 0.4.1)
- Removed allocation of SerializationContext for every message property
- Added support for possible null values in dictionary (Prepare for IncludeAllProperties)
- Ensure Cleanup never fails.
- Ensure Cleanup followed by a failed EnsureConnected doesn't lead to a broken emitter.

### CODE REVIEW

**Comment by repeatedly** [comment]
@moriyoshi Do you have a time for this patch?

**Comment by moriyoshi** [comment]
I'm taking a look into this.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 1]
```diff
@@ -24,25 +24,33 @@
...
 {
-    internal class OrdinaryDictionarySerializer: MessagePackSerializer<IDictionary<string, object>>
-    {
-        internal OrdinaryDictionarySerializer(SerializationContext ownerContext) : base(ownerContext)
-        {
-        }
-
+    internal class OrdinaryDictionarySerializer : MessagePackSerializer<IDictionary<string, object>>
+    {
+        readonly SerializationContext _embeddedContext;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: I'd prefer private instance variables not being underscored.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 2]
```diff
@@ -140,23 +157,28 @@ public void UnpackTo(Unpacker unpacker, object collection)
     internal class FluentdEmitter
     {
         private static DateTime unixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
-        private Packer packer;
-        private SerializationContext serializationContext;
+        private readonly Packer _packer;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Ditto.  Making it `readonly` looks like a plus., but what prompted you to change the names of existing fields as well as newly introduced ones?

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 3]
```diff
@@ -185,67 +207,68 @@ public class Fluentd : NLog.Targets.TargetWithLayout
 
         public bool EmitStackTraceWhenAvailable { get; set; }
 
-        private TcpClient client;
+        public bool IncludeAllProperties { get; set; }
+
+        private TcpClient _client;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Ditto.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 4]
```diff
@@ -185,67 +207,68 @@ public class Fluentd : NLog.Targets.TargetWithLayout
...
 
-        private Stream stream;
+        private Stream _stream;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Ditto.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 5]
```diff
@@ -185,67 +207,68 @@ public class Fluentd : NLog.Targets.TargetWithLayout
...
 
-        private FluentdEmitter emitter;
+        private FluentdEmitter _emitter;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Ditto.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 6]
```diff
@@ -185,67 +207,68 @@ public class Fluentd : NLog.Targets.TargetWithLayout
...
-            client.NoDelay = this.NoDelay;
-            client.ReceiveBufferSize = this.ReceiveBufferSize;
-            client.SendBufferSize = this.SendBufferSize;
-            client.SendTimeout = this.SendTimeout;
-            client.ReceiveTimeout = this.ReceiveTimeout;
-            client.LingerState = new LingerOption(this.LingerEnabled, this.LingerTime);
+        private void InitializeClient()
+        {
+            _client = new TcpClient();
+            _client.NoDelay = NoDelay;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: I don't think those need a change like this.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 7]
```diff
@@ -288,17 +311,37 @@ protected override void Write(LogEventInfo logEvent)
                 }
                 record.Add("stacktrace", transcodedFrames);
             }
-            EnsureConnected();
-            if (this.emitter != null)
+            if (IncludeAllProperties && logEvent.Properties.Count > 0)
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: My two cents, removing the shortcut check `logEvent.Propeties.Count > 0`  will not result in an obvious performance penalty.

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 8]
```diff
@@ -288,17 +311,37 @@ protected override void Write(LogEventInfo logEvent)
...
-                    this.emitter.Emit(logEvent.TimeStamp, Tag, record);
-                }
-                catch (Exception e)
+                foreach (var property in logEvent.Properties)
                 {
+                    var propertyKey = property.Key.ToString();
+                    if (string.IsNullOrEmpty(propertyKey))
+                        continue;
+
+                    record[propertyKey] = property.Value;
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Are you sure that the values of the properties are however serializable?

File: src/NLog.Targets.Fluentd/NLog.Targets.Fluentd.nuspec [hunk 9]
```diff
@@ -1,19 +0,0 @@
-<?xml version="1.0"?>
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/NLog.Targets.Fluentd.nuspec]: Do we really need this one any longer?

File: src/NLog.Targets.Fluentd/Fluentd.cs [hunk 10]
```diff
@@ -185,67 +207,68 @@ public class Fluentd : NLog.Targets.TargetWithLayout
 
         public bool EmitStackTraceWhenAvailable { get; set; }
 
-        private TcpClient client;
+        public bool IncludeAllProperties { get; set; }
```
Comment (moriyoshi) [inline src/NLog.Targets.Fluentd/Fluentd.cs]: Please document how this property works in README.md.

File: src/NLog.Targets.Fluentd/NLog.Targets.Fluentd.nuspec (hunk 9 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/NLog.Targets.Fluentd.nuspec]: The file is marked as deleted in this PR.

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 8 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Have added a guard to protect against non-trivial properties.

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 7 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: It skips the allocation of an enumerator.

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 6 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Removed underscore and added this

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 5 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Removed underscore and added this

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Removed underscore and added this

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Removed underscore and added this

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Removed underscore and added this

File: src/NLog.Targets.Fluentd/Fluentd.cs (hunk 10 above)
Comment (snakefoot) [reply src/NLog.Targets.Fluentd/Fluentd.cs]: Have updated ReadMe.md with description of IncludeAllProperties

**Comment by snakefoot** [comment]
@moriyoshi Have implemented the suggested changes.

**Comment by moriyoshi** [comment]
I'd like to keep this compatible with the older versions of NLog whose current minimum requirement of .NET version is 3.5, while .NET Standard 2.0 apparently requires 4.5 or above.  Do you have any idea on this?

**Comment by moriyoshi** [comment]
> Your current project only supports net45, so not sure why you mention net35.

Yep, but there wasn't much intention.  I'd thought the code base could be compiled with .NET 3.5 without much work at the time I published this project.

> I have configured the new Nlog.Targets.Fluentd.csproj, so Net45 uses Nlog 2.1.0, and NetStandard20 uses Nlog4.5

That sounds OK.  The problem I was seeing is that older NLog (<4.5) don't have structural logging capability, nor the accompanying API so it appears the new code base cannot be compiled with the older versions.

**Comment by snakefoot** [comment]
The structural logging is just using LogEventInfo.Properties that also exist in Nlog ver. 2.1. The Nlog.Targets.Fluentd still supports Nlog 2.1 for net45. It is only Netstandard that needs the new Nlog 4.5

I can create a new PR that adds support for net35 and net40. When this has been merged


Sent from my Samsung device


-------- Original message --------

**Comment by snakefoot** [comment]
@moriyoshi Have added net35 and net40 as build targets, so the nuget-package will also support those platforms (Also with NLog 2.1 as minimum version).

**Comment by snakefoot** [comment]
@moriyoshi Something that I'm missing, before this can be merged?

**Comment by moriyoshi** [comment]
Looks great! Apologies for leaving this for a long time.  I think I can merge this then.

**Pull request merged**